### PR TITLE
[SPARK-43291][PS] Generate proper warning on different behavior with `numeric_only`

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR added warning messages throughout the Pandas API on Spark wherever the `numeric_only` parameter is used with different default value.

This is to inform users of the different default behavior in Pandas API on Spark compared to traditional Pandas.

### Why are the changes needed?

The default behavior of the `numeric_only` parameter in Pandas API on Spark is different from Pandas 2.x.x due to the inability to support mixed data types in a single column.

This could potentially lead to unexpected results for users unfamiliar with this distinction. By providing a clear warning message, we aim to prevent confusion and potential mistakes.

### Does this PR introduce _any_ user-facing change?

Yes. Users will now see a warning message when they use functions or methods that involve the `numeric_only` parameter without explicitly setting it. This message informs them of the different default behaviors between Pandas API on Spark and traditional Pandas.

```python
>>> psdf = ps.DataFrame({"A": [1, 2, 3], "B": ['a', 'b', 'c']})
>>> psdf.cov()
/.../spark/python/pyspark/pandas/utils.py:1016: PandasAPIOnSparkAdviceWarning: In Pandas API on Spark, the 'numeric_only' parameter defaults to 'True' if not set, due to the inability to mix different data types in a single column in Spark. This behavior might differ from the traditional Pandas behavior where 'numeric_only' defaults to 'False'. Please ensure to set this parameter explicitly to avoid unexpected results.
  warnings.warn(message, PandasAPIOnSparkAdviceWarning)
     A
A  1.0
```

### How was this patch tested?

Manually tested. The existing CI should pass.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by zhengruifeng — DISMISSED** [review]

File: python/pyspark/pandas/frame.py [hunk 1]
```diff
@@ -9029,6 +9047,14 @@ def cov(self, min_periods: Optional[int] = None, ddof: int = 1) -> "DataFrame":
         """
         if not isinstance(ddof, int):
             raise TypeError("ddof must be integer")
+        if numeric_only is None:
```
Comment (zhengruifeng) [inline python/pyspark/pandas/frame.py:9050]: do we need to warn users if all columns are numeric?

File: python/pyspark/pandas/groupby.py [hunk 2]
```diff
@@ -608,18 +608,18 @@ def max(self, numeric_only: Optional[bool] = False, min_count: int = -1) -> Fram
...
         Parameters
         ----------
-        numeric_only : bool, default True
+        numeric_only : bool, default None
             Include only float, int, boolean columns. If None, will attempt to use
             everything, then use only numeric data. False is not supported.
             This parameter is mainly for pandas compatibility.
 
-            .. versionadded:: 3.4.0
+            .. versionchanged:: 4.0.0
```
Comment (zhengruifeng) [inline python/pyspark/pandas/groupby.py:622]: shall we keep the old `versionchanged` and add this new one?
both `versionchanged` should doc what was changed.

Comment (itholic) [reply python/pyspark/pandas/groupby.py:622]: Makes sense to me. Updated!

**Comment by zhengruifeng** [comment]
@itholic I think I'm missing something, I can not understand `the inability to support mixed data types in a single column`:
why we don't follow the default value (`False`)?

It look like we can check the schema and filter out non-numerical columns if needed

**Comment by itholic** [comment]
So far, we don't follow the Pandas behavior since we couldn't support the object-dtype for stat functions in some cases as beolw:
```python
# DataFrame
>>> pdf
   A  B
0  1  a
1  2  b
2  3  c

# Pandas works
>>> pdf.min(numeric_only=False)
A    1
B    a
dtype: object

# Pandas API on Spark doesn't work
>>> ps.from_pandas(pdf).min(numeric_only=False)
...
pyarrow.lib.ArrowInvalid: Could not convert 'a' with type str: tried to convert to int64
```

But on my second thought, it's a bug from our code in Pandas API on Spark so we can support `numeric_only=False` as default by fixing the existing bug.

Let me just close this ticket, and change the default value instead.

Thanks for pointing out, @zhengruifeng !

**Pull request closed without merging**